waiter/src/waiter/descriptor.clj

```diff
@@ -374,3 +376,47 @@
         (perform-authorization-checks can-run-as? auth-user descriptor)
         {:descriptor descriptor
          :latest-descriptor latest-descriptor}))))
+
+(defn await-service-goal-fallback-state-locally
+  "Polls fallback-state-atom locally and returns true if the goal state was reached and false if timeout is reached"
+  [fallback-state-atom service-id timeout sleep-duration goal]
+  (async/go
+    (let [[get-current-value-fn goal-value] (case goal
+                                              "deleted" [#(service-exists? % service-id) false]
+                                              "healthy" [#(service-healthy? % service-id) true]
+                                              "exist" [#(service-exists? % service-id) true])]
+      (loop [time-left-ms timeout]
+        (let [fallback-state @fallback-state-atom
+              current (get-current-value-fn fallback-state)
+              goal-reached? (= current goal-value)]
+          (if (or goal-reached? (not (pos? time-left-ms)))
+            goal-reached?
+            (do
+              (async/<! (async/timeout (min sleep-duration time-left-ms)))
+              (recur (- time-left-ms sleep-duration)))))))))
+
+(defn await-service-goal-fallback-state
+  "Polls local router and other routers and returns map router-id->success?"
+  [fallback-state-atom make-inter-router-requests-fn router-id service-id timeout sleep-duration goal]
+  (go-try
```

Comment:
Where can this function throw an exception?

Reply:
The function doesn't explicitly throw an exception, but this test case https://github.com/twosigma/waiter/blob/2b21e3c99b7196974e7a6aa184bf5c29c371363d/waiter/test/waiter/core_test.clj#L554
depends on the JSON error being thrown, but the error gets lost in the async/go without a try catch